### ISSUE TITLE
Model doesn't learn to classify

### ISSUE THREAD

**Issue by Bernardo-Favoreto** [microsoft/CameraTraps#150]
Hey guys!

So, I've been trying to work with this amazing repo you guys put out there but, even though I can run the scripts and execute training, the model simply doesn't seem to be learning how to classify. My goal is to train a classifier on my own species.

I will try to describe the path I made to get where I am now. 
I'm on Ubuntu 18.04, tensorflow-gpu=1.15 and have installed all necessary dependencies to run your repository.
I mainly followed this to try and achieve what I needed: https://github.com/microsoft/CameraTraps/blob/master/classification/TUTORIAL.md

First, I started with a VOC format dataset (consists of images + xml with the detections). Because here we need to use the COCO json format, I used the following script to make such changes:
https://github.com/Tony607/voc2coco/blob/master/voc2coco.py
And also, here's an example of an original XML file.
![image](https://user-images.githubusercontent.com/32913431/78784280-70491f00-797b-11ea-8254-7b4914568388.png)
Now I bring up the question: the "path" field is not correct, as this file is not in the original place anymore. However, it doesn't seem to affect anything, because I can still generate the COCO json. Here's an example extracted from said json:
![image](https://user-images.githubusercontent.com/32913431/78785346-645e5c80-797d-11ea-8cb1-b9d07546fb67.png)
Which seems ok, since it got the file name right (this is the name it'll search when I point to the images folder, right?) and as I understand COCO json doesn't use **xmax and ymax** but rather width and height, right as well? The category id also matches the one expected. 
Alright, the next step was generating the tf records. 
I started trying by using your script, but it seemed to be trying to import a file that did not exist, so instead I went in and used this one (from TF OD API):
https://github.com/tensorflow/models/blob/master/research/object_detection/dataset_tools/create_coco_tf_record.py
As for the test and validation folders, I pointed it to the same and simply ignored the validation files (since here only train and test matters). Everything went well and I got my folder full of tf records (I just had to rename them because the training script here searches for "train-" and "test-" so I add these to the beginning of my file name.
So, with everything in hand, I went to the training part.
I had already the tf records and model checkpoint (inception v4), just needed to prepare some files for training. First copy and modify the dataset python script. I changed the "SPLIT_TO_SIZE" variable to:  SPLITS_TO_SIZES = {'train': 7624, 'test': 1885} and "_NUM_CLASSES = 17". Furthermore, I added my dataset to "dataset_factory.py". Finally, I modify the .sh script to adapt and create the one for my own dataset.
DATASET_DIR points to the folder containing the tf records. TRAIN_DIR to the output path (log/date...). CHECKPOINT_PATH to where I had my inception model checkpoint. And MODEL_NAME was still "inception_v4". For better visualization:
![image](https://user-images.githubusercontent.com/32913431/78787159-49411c00-7980-11ea-917e-a0a8acd25ea9.png)

Finally, I executed the script by typing: bash ../training_scripts/train_deep_inception_v4.sh
It trained reeeeally slowly, with the loss decreasing .0001 per 300~400 steps. It seemed ok though cause it was actually going down. I left the model training overnight, unfortunately it trained only for a few hours. 
However, when I woke up and came to check the model, it had stopped immediately after an evaluation step which outputed accuracy and recall equal to 1, which I knew wasn't right. I then exported the graph to see an actual inference and this is what I got:
![image](https://user-images.githubusercontent.com/32913431/78787483-ccfb0880-7980-11ea-9703-f96a2b93089d.png)
So, the detection is working well, since I use the megadetector model you guys provide, however, classification absolutely sucks. What it is doing is simply labeling all animals as the first label it sees on my "label.txt" file. If I change which class comes first, classification changes as well. So, it doesn't like the model is learning to classify anything, and I simply can't find a reason why. 
I'm working on this problem for quite a while now, have tried to use TF OD API directly but with no success as well.

Any tips? Directions? Questions?
Please any help is welcome!

**Comment by yangsiyu007** [comment]
Hi there! The final evaluation step outputting accuracy and recall of 1 makes me suspicious of the input tfrecords that you made for training the classifier. Did you inspect them to see if they had the correct species/class labels? You can use something like this [notebook](https://github.com/microsoft/CameraTraps/blob/master/data_management/tfrecords/tools/inspect_tfrecords.ipynb) to check.

**Comment by Bernardo-Favoreto** [comment]
Hey, thanks for answering!

I actually did use this notebook to do exactly that, and everything is apparently alright. Here's an example of a random tf record (with no image feature cause that's just uncomprehensible):
**_feature {
    key: "image/filename"
    value {
      bytes_list {
        value: "09160182.AVI_90.png"
      }
    }
  }
  feature {
    key: "image/format"
    value {
      bytes_list {
        value: "jpeg"
      }
    }
  }
  feature {
    key: "image/height"
    value {
      int64_list {
        value: 720
      }
    }
  }
  feature {
    key: "image/key/sha256"
    value {
      bytes_list {
        value: "90390f8a7cdf472524dafb9701e091079dbba4b0d4ba2fce5cc6eb037a7c4eb5"
      }
    }
  }
  feature {
    key: "image/object/area"
    value {
      float_list {
        value: 51302.0
      }
    }
  }
  feature {
    key: "image/object/bbox/xmax"
    value {
      float_list {
        value: 0.18281249701976776
      }
    }
  }
  feature {
    key: "image/object/bbox/xmin"
    value {
      float_list {
        value: 0.0062500000931322575
      }
    }
  }
  feature {
    key: "image/object/bbox/ymax"
    value {
      float_list {
        value: 0.7902777791023254
      }
    }
  }
  feature {
    key: "image/object/bbox/ymin"
    value {
      float_list {
        value: 0.4749999940395355
      }
    }
  }
  feature {
    key: "image/object/class/text"
    value {
      bytes_list {
        value: "Veado"
      }
    }
  }
  feature {
    key: "image/object/is_crowd"
    value {
      int64_list {
        value: 0
      }
    }
  }
  feature {
    key: "image/source_id"
    value {
      bytes_list {
        value: "7"
      }
    }
  }
  feature {
    key: "image/width"
    value {
      int64_list {
        value: 1280
      }
    }
  }
}_**

What I think is slightly weird is that image format is said to be jpeg but the file is actually .png. Something else that came to mind, was that the next cell it's used to show the image. When it tries to find the image using "path", it is unable to because the tf record is actually holding only the file name, not an absolute path (as the attribute suggests: image/filename and not image/path):
![image](https://user-images.githubusercontent.com/32913431/78898610-75c16a80-7a4a-11ea-95ec-eed477b6041c.png)

I hope this can help clarify things. Please feel free to ask more questions if it's still unclear!

Thank you!

**Comment by yangsiyu007** [comment]
Okay, looks like the class label is on the object, not at the image level. Taking one step back: would you like to train an object detector that classifies each detection by its species, i.e. a multi-species detector? 

The MegaDetector only classifies its detections as `animal` or `person`; our idea (the idea of the `classification` directory) is to use the MegaDetector to find animals, crop out the predicted bounding box with some padding, and then just train an *image* classifier for these crop-outs.

**Comment by agentmorris** [comment]
Closing this issue for lack of response to the previous question; please feel free to comment here or to contact us at cameratraps@microsoft.com if you want to discuss this further.  Thanks!
